deeplearning4j/deeplearning4j-cuda/src/main/java/org/deeplearning4j/nn/layers/dropout/CudnnDropoutHelper.java

```diff
@@ -0,0 +1,147 @@
+        private cudnn.cudnnTensorStruct yTensorDesc = new cudnn.cudnnTensorStruct();    //Output
+        private cudnn.cudnnTensorStruct dyTensorDesc = new cudnn.cudnnTensorStruct();   //Grad at output
+        private cudnn.cudnnDropoutStruct dropoutDesc = new cudnn.cudnnDropoutStruct();
+
+        public CudnnDropoutContext() {
+            createHandles();
+            deallocator(new Deallocator(this));
+        }
+
+        public CudnnDropoutContext(CudnnDropoutContext c) {
+            super(c);
+            xTensorDesc = new cudnn.cudnnTensorStruct(c.xTensorDesc);
+            dxTensorDesc = new cudnn.cudnnTensorStruct(c.dxTensorDesc);
+            yTensorDesc = new cudnn.cudnnTensorStruct(c.yTensorDesc);
+            dyTensorDesc = new cudnn.cudnnTensorStruct(c.dyTensorDesc);
+            dropoutDesc = new cudnn.cudnnDropoutStruct(c.dropoutDesc);
+        }
+
+        @Override
+        protected void createHandles() {
+            super.createHandles();
+            checkCudnn(cudnnCreateTensorDescriptor(xTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(dxTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(yTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(dyTensorDesc));
+            checkCudnn(cudnnCreateDropoutDescriptor(dropoutDesc));
+        }
+
+        @Override
+        protected void destroyHandles() {
+            checkCudnn(cudnnDestroyTensorDescriptor(xTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(dxTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(yTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(dyTensorDesc));
+            checkCudnn(cudnnDestroyDropoutDescriptor(dropoutDesc));
+            super.destroyHandles();
+        }
+    }
+
+    private CudnnDropoutContext cudnnContext = new CudnnDropoutContext();
+    private Pointer statesPtr;
+    private long stateSizeBytes;
+    private long seed;
+
+
+    @Override
+    public void applyDropout(INDArray input, INDArray resultArray, double dropoutInputRetainProb) {
+        float p = (float)(1.0 - dropoutInputRetainProb);    //CuDNN uses p = probability of setting to 0
+
+        checkCudnn(cudnnSetTensorNdDescriptor(cudnnContext.xTensorDesc, dataType, input.rank(),
+                //TODO int cast
+                ArrayUtil.toInts(input.shape()),
+                ArrayUtil.toInts(input.stride())));
+        checkCudnn(cudnnSetTensorNdDescriptor(cudnnContext.yTensorDesc, dataType, resultArray.rank(),
+                //TODO int cast
+                ArrayUtil.toInts(resultArray.shape()),
+                ArrayUtil.toInts(resultArray.stride())));
+
+        //Reserve space
+        SizeTPointer stateSizeBytesPtr = new SizeTPointer();        //TODO correct???
+        stateSizeBytes = stateSizeBytesPtr.get();
+        checkCudnn(cudnnDropoutGetReserveSpaceSize(cudnnContext.xTensorDesc, stateSizeBytesPtr));
```

Comment:
We'd need to call `stateSizeBytesPtr.get()` after `cudnnDropoutGetReserveSpaceSize()` for this to make sense.